### PR TITLE
[iOS] allows default text renderers to be used in their custom versio…

### LINKED ISSUE #3456: [iOS][Extending renderer] [TextBlock renderer crashes when extended]
# Platform

What platform is your issue or question related to? (Delete other platforms).

* iOS

# Author or host

Author/host, Premera app

# Version of SDK

1.2.2

# Details

I am trying to add custom styling to card elements. For action elements, your examples indicate I need to call your render function from mine, and then I can apply my custom styling to the generated UIKit controls. However, if I do the same for static elements your render function crashes. This is due to logic in ACRView.addTasksToConcurrentQueue which decides to bypass logic that the stock renderer relies on, whenever the renderer is overridden.

To work around this, I have to implement a render function in Objective C++ and implement the renderer in its entirety, the exact opposite of the pattern for an action element. This also limits the pool of developers who can work on this (many newer developers have only used Swift), and making adoption of Adaptive cards less attractive to iOS developers.

### PR DESCRIPTION
**Pull request by nesalang**
## Related Issue
fixes #3456 

## Description
when text renderers are overridden, but they are used in their overridden renderers, the crashes were observed because the default renderers need intermediate text processing result.

this change refactored intermediate text processing into a function. Change added the function's function call to default renderers, and it will be used when they are overridden.

## How Verified
1. units tests were added.
2. tested with custom text block renderer with sample app

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/3498)

### CODE REVIEW

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizerTests/ADCIOSVisualizerTests.mm [hunk 1]
```diff
@@ -108,6 +111,49 @@ - (void)testACRTextView
...
+{
+    ACRRegistration *registration = [ACRRegistration getInstance];
+    [registration setBaseCardElementRenderer:[ACRTextBlockRenderer getInstance] cardElementType:ACRCardElementType::ACRTextBlock];
+
+    NSString *payload = [NSString stringWithContentsOfFile:[_mainBundle pathForResource:@"Feedback" ofType:@"json"] encoding:NSUTF8StringEncoding error:nil];
+    ACOAdaptiveCardParseResult *cardParseResult = [ACOAdaptiveCard fromJson:payload];
+    XCTAssertTrue(cardParseResult && cardParseResult.isValid);
+    ACRRenderResult *renderResult = [ACRRenderer render:cardParseResult.card config:_defaultHostConfig widthConstraint:335];
+
+    XCTAssertNotNil(renderResult.view);
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizerTests/ADCIOSVisualizerTests.mm]: looks like this pattern is used a lot -- helper function?

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizerTests/ADCIOSVisualizerTests.mm [hunk 2]
```diff
@@ -108,6 +111,49 @@ - (void)testACRTextView
     XCTAssertTrue([acrTextView.text length] == 0);
 }
 
+// this test ensure that having custom text render doesn't crash
+// in use case where custom renderer uses default text renderer
+- (void)testCustomTextRendereDoesNotCrash
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizerTests/ADCIOSVisualizerTests.mm]: >Rendere [](start = 22, length = 7)

nitpick: should be `Renderer`

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizerTests/ADCIOSVisualizerTests.mm [hunk 3]
```diff
@@ -108,6 +111,28 @@ - (void)testACRTextView
...
+    [registration setBaseCardElementRenderer:[ACRRichTextBlockRenderer getInstance] cardElementType:ACRCardElementType::ACRRichTextBlock];
+    [registration setBaseCardElementRenderer:[ACRFactSetRenderer getInstance] cardElementType:ACRCardElementType::ACRFactSet];
+
+    // TextBlock.Maxlines is used for testing when text block renderer is overriden
+    // RichTextBlock tests for RichTextBlock renderer extension
+    // ActivityUpdate tests TextBlock & FactSet extension combinations
+    NSArray<NSString *> *payloadNames = @[ @"TextBlock.MaxLines", @"RichTextBlock", @"ActivityUpdate" ];
+
+    NSArray<ACOAdaptiveCard *> *cards = [self prepCards:payloadNames];
+    // Put the code you want to measure the time of here.
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizerTests/ADCIOSVisualizerTests.mm]: >// Put the code you want to measure the time of here. [](start = 4, length = 53)

I don't think this comment belongs here.

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRFactSetRenderer.mm [hunk 4]
```diff
@@ -123,16 +124,25 @@ - (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
 
     [factSetWrapperView adjustHuggingForLastElement];
 
+    BOOL isOverridden = [[ACRRegistration getInstance] isElementRendererOverridden:ACRCardElementType::ACRFactSet];
+
     for (auto fact : fctSet->GetFacts()) {
         NSString *title = [NSString stringWithCString:fact->GetTitle().c_str() encoding:NSUTF8StringEncoding];
+        NSString *titileElemId = [key stringByAppendingString:[[NSNumber numberWithInt:rowFactId++] stringValue]];
```
Comment (bviglietta) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRFactSetRenderer.mm]: >titile [](start = 18, length = 6)

`title`

**Pull request merged**